File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlStatusDot.kt
package com.dizyx.nockerl.design.components

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.NockerlColors
import com.dizyx.nockerl.design.tokens.NockerlMotionDuration

/**
 * The standardized **status dot**: a small filled disc whose hue names a state,
 * optionally pulsing. Unifies the raw `Box + PulsingDot` pattern client #1
 * re-derives at four call sites into one wrapper (status-dot.mdx).
 *
 * Two hue ladders, mirroring the shipped Swift `NockerlStatusDot` exactly:
 * - the semantic ladder ([NockerlStatusDotStatus] → status tokens), and
 * - a raw palette color (the session-dot idiom: pass [NockerlColors.dotStreaming],
 *   [NockerlColors.dotUnread], …, but never a literal).
 *
 * Pulse (design-laws §7): animates ALPHA only (1.0 ↔ 0.3) at the RATIFIED 800ms
 * motion value (r2, B6; [tween] + [RepeatMode.Reverse]). The web-only "ping" ring
 * is DROPPED by the same ruling. No platform implements it.
 *
 * A11y (law §13): state is never color-alone; pass a [contentDescription] (or use
 * the semantic overload, which derives one from the status name).
 */
@Composable
fun NockerlStatusDot(
    color: Color,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    size: Dp = NockerlStatusDotDefaults.LoudSize,
    pulse: Boolean = false,
) {
    val alpha =
        if (pulse) {
            val transition = rememberInfiniteTransition(label = "nockerlStatusDot")
            val pulsing by transition.animateFloat(
                initialValue = 1f,
                targetValue = NockerlStatusDotDefaults.PULSE_FLOOR_ALPHA,
                animationSpec =
                    infiniteRepeatable(
                        animation = tween(durationMillis = NockerlStatusDotDefaults.PULSE_MS),
                        repeatMode = RepeatMode.Reverse,
                    ),
                label = "nockerlStatusDotAlpha",
            )
            pulsing
        } else {
            1f
        }

    // Captured under a distinct name: inside the semantics lambda the receiver's own
    // `contentDescription` property shadows the parameter (the classic apply-style
    // self-assignment trap: reading that property throws at runtime).
    val accessibleName = contentDescription
    val semantics =
        if (accessibleName != null) {
            Modifier.semantics { this.contentDescription = accessibleName }
        } else {
            Modifier
        }

    Box(
        modifier =
            modifier
                .then(semantics)
                .size(size)
                .alpha(alpha)
                .clip(CircleShape)
                .background(color),
    )
}

/**
 * Semantic-ladder overload: hue AND the default accessible name come from [status].
 *
 * @param status the ladder step.
 * @param modifier outer modifier.
 * @param contentDescription override the derived accessible name.
 * @param size disc diameter: [NockerlStatusDotDefaults.LoudSize] (8dp, default) or
 *   [NockerlStatusDotDefaults.QuietSize] (6dp, ambient).
 * @param pulse animate the live-state alpha pulse.
 */
@Composable
fun NockerlStatusDot(
    status: NockerlStatusDotStatus,
    modifier: Modifier = Modifier,
    contentDescription: String? = null,
    size: Dp = NockerlStatusDotDefaults.LoudSize,
    pulse: Boolean = false,
) {
    NockerlStatusDot(
        color = status.color(LocalNockerlColors.current),
        contentDescription = contentDescription ?: status.accessibleName,
        modifier = modifier,
        size = size,
        pulse = pulse,
    )
}

/** The semantic status ladder (mirrors the Swift `NockerlStatusDotStatus` 1:1). */
enum class NockerlStatusDotStatus(
    /** The default accessible name when the caller supplies none. */
    val accessibleName: String,
) {
    /** Healthy / complete (`statusSuccess`). */
    SUCCESS("Success"),

    /** Needs attention (`statusWarning`). */
    WARNING("Warning"),

    /** Failed / blocked (`statusError`). */
    ERROR("Error"),

    /** Informational (`statusInfo`). */
    INFO("Info"),

    /** Inactive / unknown (`dotIdle`). */
    NEUTRAL("Neutral"),
    ;

    /** Resolve the ladder step to its palette slot. */
    fun color(colors: NockerlColors): Color =
        when (this) {
            SUCCESS -> colors.statusSuccess
            WARNING -> colors.statusWarning
            ERROR -> colors.statusError
            INFO -> colors.statusInfo
            NEUTRAL -> colors.dotIdle
        }
}

/** Shared status-dot constants (sizes + the pinned pulse contract). */
object NockerlStatusDotDefaults {
    /** Loud 8dp disc: live / attention states (the default). */
    val LoudSize: Dp = 8.dp

    /** Quiet 6dp disc: idle / ambient states. */
    val QuietSize: Dp = 6.dp

    /** Pulse alpha floor (1.0 → 0.3, interpolatable only, law §7). */
    const val PULSE_FLOOR_ALPHA = 0.3f

    /** Pulse half-period: the `motion.duration.pulse` TOKEN (B6 / r3 emitted). */
    val PULSE_MS = NockerlMotionDuration.pulseMs
}

File: Sources/NockerlDesign/Components/NockerlStatusDot.swift
// NockerlStatusDot: the semantic status-dot ladder for SwiftUI surfaces.
//
// Mirrors the ratified status dot (status-dot.mdx): a small filled disc whose hue names
// a state, with an optional visible label and an optional live PULSE. Two hue ladders:
//   - the semantic status ladder (success / warning / error / info / neutral), and
//   - a raw palette color (for the session-dot tokens: dotStreaming, dotAttention, …).
//
// Pulse (law §7): animates OPACITY only (1.0 ↔ 0.3), never a fill swap, and freezes
// under Reduce Motion. Duration is the RATIFIED 800ms motion value (Voice's
// old 600ms recording dot conforms at swap). The web-only "ping" ring is DROPPED by
// the same ruling: no platform implements it.
//
// A11y (law §13): state is never color-alone. The accessible name is the visible
// label when present, otherwise the status's own name.

import SwiftUI

/// The semantic status ladder for ``NockerlStatusDot``.
public enum NockerlStatusDotStatus: Equatable {
    /// Healthy / complete (`statusSuccess`).
    case success
    /// Needs attention (`statusWarning`).
    case warning
    /// Failed / blocked (`statusError`).
    case error
    /// Informational (`statusInfo`).
    case info
    /// Inactive / unknown (`dotIdle`).
    case neutral

    /// Resolve the ladder step to its palette color.
    func color(in palette: NockerlPalette) -> Color {
        switch self {
        case .success: return palette.statusSuccess
        case .warning: return palette.statusWarning
        case .error: return palette.statusError
        case .info: return palette.statusInfo
        case .neutral: return palette.dotIdle
        }
    }

    /// The default accessible name when no visible label is supplied.
    var accessibleName: String {
        switch self {
        case .success: return "Success"
        case .warning: return "Warning"
        case .error: return "Error"
        case .info: return "Info"
        case .neutral: return "Neutral"
        }
    }
}

/// The size ramp: quiet (6pt, ambient) or loud (8pt, demands the eye).
public enum NockerlStatusDotSize: Equatable {
    /// 6pt for idle / ambient states.
    case quiet
    /// 8pt for live / attention states (the default).
    case loud

    /// The disc diameter.
    var diameter: CGFloat {
        switch self {
        case .quiet: return 6
        case .loud: return NockerlSpace.space2
        }
    }
}

/// A semantic status dot: a filled disc + optional label, optionally pulsing.
public struct NockerlStatusDot: View {
    /// How the dot's hue is chosen.
    private enum Hue {
        case status(NockerlStatusDotStatus)
        case raw(Color)

        func color(in palette: NockerlPalette) -> Color {
            switch self {
            case .status(let status): return status.color(in: palette)
            case .raw(let color): return color
            }
        }
    }

    private let hue: Hue
    private let label: String?
    private let pulse: Bool
    private let size: NockerlStatusDotSize
    private let accessibleName: String

    /// Pulse opacity floor (1.0 → 0.3, mirroring the Compose `PulsingDot`).
    static let pulseFloorOpacity: Double = 0.3
    /// Pulse half-period: the `motion.duration.pulse` TOKEN (r2 B6 / r3 emitted).
    static let pulseDuration: TimeInterval = NockerlMotionDuration.pulse

    /// A dot on the semantic status ladder.
    /// - Parameters:
    ///   - status: the ladder step (drives hue AND the default accessible name).
    ///   - label: optional visible label beside the dot.
    ///   - pulse: animate the live-state opacity pulse.
    ///   - size: quiet 6pt or loud 8pt (default loud).
    public init(
        status: NockerlStatusDotStatus,
        label: String? = nil,
        pulse: Bool = false,
        size: NockerlStatusDotSize = .loud
    ) {
        let name = label.flatMap(NockerlA11y.accessibleName(from:))
        self.hue = .status(status)
        self.label = name
        self.pulse = pulse
        self.size = size
        self.accessibleName = name ?? status.accessibleName
    }

    /// A dot with an explicit palette color, the session-dot idiom
    /// (`palette.dotStreaming`, `palette.dotUnread`, …).
    /// - Parameters:
    ///   - color: the dot hue. Pass a palette token, never a literal.
    ///   - label: the REQUIRED accessible name (a raw-color dot has no status to
    ///     derive a name from; unnamed state is the a11y breach, law §13).
    ///   - pulse: animate the live-state opacity pulse.
    ///   - size: quiet 6pt or loud 8pt (default loud).
    public init(
        color: Color,
        label: String,
        pulse: Bool = false,
        size: NockerlStatusDotSize = .loud
    ) {
        let name = NockerlA11y.accessibleName(from: label)
        assert(
            name != nil,
            "NockerlStatusDot(color:) requires a non-empty accessible name. A raw-color dot is unnamed state otherwise (design-laws §13)."
        )
        self.hue = .raw(color)
        self.label = name
        self.pulse = pulse
        self.size = size
        self.accessibleName = name ?? ""
    }

    public var body: some View {
        DotBody(hue: hue, label: label, pulse: pulse, size: size)
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(accessibleName)
    }

    /// Environment-reading render body: palette + reduce-motion live here.
    private struct DotBody: View {
        let hue: Hue
        let label: String?
        let pulse: Bool
        let size: NockerlStatusDotSize

        @Environment(\.colorScheme) private var colorScheme
        @Environment(\.accessibilityReduceMotion) private var reduceMotion
        @State private var dimmed = false

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            let color = hue.color(in: palette)
            // Pulse = opacity only (law §7); frozen at full opacity under Reduce
            // Motion so the state stays legible without decoration-pulsing.
            let animating = pulse && !reduceMotion

            HStack(spacing: NockerlSpace.space2) {
                Circle()
                    .fill(color)
                    .frame(width: size.diameter, height: size.diameter)
                    .opacity(animating && dimmed ? NockerlStatusDot.pulseFloorOpacity : 1)
                    .animation(
                        animating
                            ? .easeInOut(duration: NockerlStatusDot.pulseDuration)
                                .repeatForever(autoreverses: true)
                            : nil,
                        value: dimmed
                    )
                    .onAppear {
                        if animating { dimmed = true }
                    }

                if let label {
                    Text(label)
                        .font(.nockerl(size: NockerlFontSize.size12, weight: .light))
                        .foregroundColor(palette.onCardMuted)
                }
            }
        }
    }
}
